astropy/coordinates/builtin_frames/icrs_cirs_transforms.py

```diff
@@ -186,3 +188,55 @@ def gcrs_to_gcrs(from_coo, to_frame):
     else:
         # like CIRS, we do this self-transform via ICRS
         return from_coo.transform_to(ICRS).transform_to(to_frame)
+
+@frame_transform_graph.transform(FunctionTransform, GCRS, Heliocentric)
+def gcrs_to_hcrs(gcrs_coo, hcrs_frame):
+
+    if np.any(gcrs_coo.obstime != hcrs_frame.obstime):
+        # if they GCRS obstime and HCRS obstime are not the same, we first
+        # have to move to a GCRS where they are.
+        frameattrs = gcrs_coo.get_frame_attr_names()
+        frameattrs['obstime'] = hcrs_frame.obstime
+        gcrs_coo = gcrs_coo.transform_to(GCRS(**frameattrs))
+        
+    srepr = gcrs_coo.represent_as(UnitSphericalRepresentation)
+    gcrs_ra = srepr.lon.to(u.radian).value
+    gcrs_dec = srepr.lat.to(u.radian).value
+
+    # set up the astrometry context for ICRS<->GCRS and then convert to ICRS
+    # coordinate direction
+    pv = np.array([gcrs_coo.obsgeoloc.value,
+                   gcrs_coo.obsgeovel.value])
+    tdb = hcrs_frame.obstime.tdb
+    astrom = erfa.apcs13(tdb.jd1, tdb.jd2, pv)
+    i_ra, i_dec = erfa.aticq(gcrs_ra, gcrs_dec, astrom)
+
+    if gcrs_coo.data.get_name() == 'unitspherical'  or gcrs_coo.data.to_cartesian().x.unit == u.one:
```

Comment:
The code you suggest doesn't work - ```gcrs_coo.data.components``` gives a tuple of strings - ```(lat, lon)``` for example. I don't think it's that expensive either, since the cartesian representation is only calculated if the first condition is not true, which I think it always should be.